### ISSUE TITLE
Rule S2259: False postive on typeof(T)

### ISSUE THREAD

**Issue by Corniel** [SonarSource/sonar-csharp#1326]
### Description
Rule S2259 does not regonize the assignment of a variable, based on typeof(T)

### Repro steps
``` C#
void Test<T>()
{
    var model = typeof(T);
    var prop = model.GetProperty("Value", BindingFlags.Public | BindingFlags.Instance); // Compliant
}

**Comment by Evangelink** [comment]
Took the liberty to edit your post just to remove an extra `5` in the rule ID.

**Comment by Evangelink** [comment]
Hi @Corniel,

I have tried your example on a ConsoleApplication with `SonarAnalyzer.CSharp 6.8.2.4717` (latest NuGet version released) and couldn't reproduced the issue. Could you tell me more?

**Comment by Corniel** [comment]
Me bad @Evangelink . I thought lets make the snappet code a simple as possible, without testing if it was not over simplificering:

### Repro steps
``` C#
void Test<T>()
{
    var model = typeof(T);
    PropertyInfo prop = null;

  //while(true)
   for(var i = 0; i < 10; i++)
   {
        prop = model.GetProperty("Value", BindingFlags.Public | BindingFlags.Instance); // Compliant
   }
}
```

So it complains due to the loop and the once assigned as null. (You could argue that is code is smelly, and I think that is true. Howevr, the model should  never be null.

**Comment by Corniel** [comment]
@Evangelink Did you give my updated example a try?

**Comment by Evangelink** [comment]
Hi @Corniel,

I think that your example is still having some issue and so I don't manage to have S2259 raised on `prop = model.Get....`. Looking at the implementation I see that we apply a not null constraint when we see `model = typeof(T)` because it doesn't ever return null.

**Comment by agigleux** [comment]
Hi @Corniel ,

Do you still have the issue with SonarC# 7.1?
Can you provide another reproducer because so far we did not manage to reproduce your problem with the provided code snippet?

Thanks

**Comment by Evangelink** [comment]
Hi @Corniel,

Sorry to ping you but I'd like to know if you are still facing the issue and if so if you could provide another code snippet. It seems like the 2 use cases you provided are not causing any issue on our side.

**Comment by Corniel** [comment]
Sorry Guys, forgot to take follow up actions on this. I'll try to reproduce the issue this week.

**Comment by Corniel** [comment]
I cannot find back the code that lead to this error. :(

Sorry for the inconvenience.
